async-book/src/execution/wakeups.md

```diff
@@ -0,0 +1,135 @@
+# Task Wakeups with `LocalWaker` and `Waker`
+
+It's common that futures aren't able to complete the first time they are
+`poll`ed. When this happens, the future needs to ensure that it is polled
+again once it is ready to make more progress. This is done with the
+`LocalWaker` and `Waker` types.
+
+`LocalWaker` and `Waker` each provide a `wake()` method that can be used to
+tell the executor that the appropriate task should be awoken, and its top-level
+future should be polled again. Both types are also implement `clone()` so that
+they can be copied around and stored. The difference between the two is
+thread-safety: `LocalWaker` is `!Send` and `!Sync`, and so cannot be used from
+threads other than the one it was created from. This allows `LocalWaker`
+implementations to perform special optimized behavior for the current thread.
+`Waker`s, on the other hand, are `Send` and `Sync`, and so can be used across
+multiple threads. A `LocalWaker` can be turned into a thread-safe `Waker` using
+the `into_waker()` function. This function is free to call-- it doesn't
+allocate at runtime or anything similar, but calling `wake()` on the resulting
+`Waker` may be less performant than calling `wake()` on the original
+`LocalWaker`.
+
+Let's try implementing a simple timer future using `Waker` and `LocalWaker`.
+
+## Applied: Build a Timer
+
+For the sake of the example, we'll just spin up a new thread when the timer
```

Comment:
Yes, an executor is necessary, but building an executor is more complicated than building a single future. When I taught this as a class, I had a basic pre-made executor that I provided to the group. However, including pre-made code into a book is somewhat awkward. Perhaps we should bring in futures-rs here and use `block_on` in order to demonstrate?

Reply:
I think for a book that would be perfectly executable. Say that that executors are required, so we'll bring in futures for block_on, and the reader will learn more about executors in chapter X.

Reply:
Looking back at this, the very next section is on executors, and it gives an example using `TimerFuture` at the end. It doesn't seem unreasonable to have users wait for the end of the next section to run the code, especially since it was all pre-provided.